File: java-samples/src/main/java/posts/parthmistry/javasamples/MapTest.java
package posts.parthmistry.javasamples;

import posts.parthmistry.javasamples.utils.ElapsedTimeMonitor;

import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.HashMap;

public class MapTest {

    public static void main(String[] args) throws Exception {
        var uuidFilePath1 = args[0];
        var uuidFilePath2 = args[1];

        var uuidList = new ArrayList<String>();
        var uuidMap = new HashMap<String, Integer>();

        try (var uuidLines = Files.lines(Paths.get(uuidFilePath1))) {
            uuidLines.forEach(uuidList::add);
        }

        try (var uuidLines = Files.lines(Paths.get(uuidFilePath2))) {
            uuidLines.forEach(uuid -> uuidMap.put(uuid, 1));
        }

        var iterations = 20;

        var globalTimeMonitor = new ElapsedTimeMonitor();
        var iterationDurationSum = 0L;

        for (int i = 0; i < iterations; ++i) {
            var iterationTimeMonitor = new ElapsedTimeMonitor();
            var iterationSum = 0L;

            for (var uuid : uuidList) {
                var mapValue = uuidMap.get(uuid);
                if (mapValue != null) {
                    iterationSum += mapValue;
                }
            }

            var iterationDuration = iterationTimeMonitor.getElapsedTimeMillis();
            iterationDurationSum += iterationDuration;

            System.out.println("iteration sum: " + iterationSum);
            System.out.println("iteration duration: " + iterationDuration + " ms");
        }

        System.out.println("average iteration duration: " + (iterationDurationSum / ((double) iterations)) + " ms");
        System.out.println("global duration: " + globalTimeMonitor.getElapsedTimeMillis() + " ms");
    }

}

File: python-samples/map_test.py
import sys
from utils import ElapsedTimeMonitor

uuid_file_path1 = sys.argv[1]
uuid_file_path2 = sys.argv[2]

uuid_list = []
uuid_map = dict()

with open(uuid_file_path1) as file:
    while line := file.readline():
        uuid_list.append(line.rstrip('\n'))

with open(uuid_file_path2) as file:
    while line := file.readline():
        uuid_map[line.rstrip('\n')] = 1

iterations = 20

global_time_monitor = ElapsedTimeMonitor()
iteration_duration_sum = 0

for n in range(0, iterations):
    iteration_time_monitor = ElapsedTimeMonitor()
    iteration_sum = 0

    for uuid in uuid_list:
        map_value = uuid_map[uuid]
        if map_value is not None:
            iteration_sum += map_value

    iteration_duration = iteration_time_monitor.get_elapsed_time_millis()
    iteration_duration_sum += iteration_duration

    print("iteration sum:", iteration_sum)
    print("iteration duration:", iteration_duration, "ms")

print("average iteration duration:", (iteration_duration_sum / float(iterations)), "ms")
print("global duration:", global_time_monitor.get_elapsed_time_millis(), "ms")
